diff --git a/include/boost/ut.hpp b/include/boost/ut.hpp
index b39e964..828cfbf 100644
--- a/include/boost/ut.hpp
+++ b/include/boost/ut.hpp
@@ -1020,8 +1020,8 @@ struct eq_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1047,8 +1047,8 @@ struct approx_ : op {
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
   [[nodiscard]] constexpr auto epsilon() const { return get(epsilon_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const TEpsilon epsilon_{};
   const bool value_{};
 };
@@ -1085,8 +1085,8 @@ struct neq_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1108,8 +1108,8 @@ struct gt_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1131,8 +1131,8 @@ struct ge_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1162,8 +1162,8 @@ struct lt_ : op {
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
  private:
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1185,8 +1185,8 @@ struct le_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1201,8 +1201,8 @@ struct and_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1217,8 +1217,8 @@ struct or_ : op {
   [[nodiscard]] constexpr auto lhs() const { return get(lhs_); }
   [[nodiscard]] constexpr auto rhs() const { return get(rhs_); }
 
-  const TLhs lhs_{};
-  const TRhs rhs_{};
+  const TLhs& lhs_{};
+  const TRhs& rhs_{};
   const bool value_{};
 };
 
@@ -1230,7 +1230,7 @@ struct not_ : op {
   [[nodiscard]] constexpr operator bool() const { return value_; }
   [[nodiscard]] constexpr auto value() const { return get(t_); }
 
-  const T t_{};
+  const T& t_{};
   const bool value_{};
 };
 
@@ -2410,7 +2410,7 @@ struct that_ {
       return static_cast<bool>(t_);
     }
 
-    const T t_{};
+    const T& t_{};
   };
 
   template <class T>
@@ -2771,7 +2771,7 @@ template <class F, class T>
 
 template <class F, template <class...> class T, class... Ts>
   requires(!std::ranges::range<T<Ts...>>)
-[[nodiscard]] constexpr auto operator|(const F& f, const T<Ts...>& t) {
+[[nodiscard]] constexpr auto operator|(const F& f, T<Ts...>&& t) {
   constexpr auto unique_name = []<class TArg>(std::string_view name,
                                               const TArg& arg, int& counter) {
     auto ret = std::string{name} + " (";
@@ -2783,20 +2783,20 @@ template <class F, template <class...> class T, class... Ts>
     return ret;
   };
 
-  return [f, t, unique_name](std::string_view type, std::string_view name) {
+  return [f, &t, unique_name](std::string_view type, std::string_view name) {
     int counter = 1;
     apply(
-        [=, &counter](const auto&... args) {
+        [=, &counter](auto&&... args) {
           (detail::on<F>(events::test<F, Ts>{
                .type = type,
                .name = unique_name.template operator()<Ts>(name, args, counter),
                .tag = {},
                .location = {},
-               .arg = args,
+               .arg = std::move(args),
                .run = f}),
            ...);
         },
-        t);
+        std::move(t));
   };
 }
 
